src/libexpr/flake/lockfile.cc

```diff
@@ -45,16 +45,26 @@ StorePath LockedNode::computeStorePath(Store & store) const
     return lockedRef.input.computeStorePath(store);
 }
 
-std::shared_ptr<Node> LockFile::findInput(const InputPath & path)
-{
+
+static std::shared_ptr<Node> doFind(const ref<Node>& root, const InputPath & path, std::vector<std::string>& visited) {
     auto pos = root;
 
+    auto pathS = printInputPath(path);
```

Comment:
Do we need to eagerly print the paths? Couldn't we have `visited` be a `std::vector<FlakeId>` and only print the paths if we get a cycle?

(not that it matters too much from a performance pov, but it would make the intent clearer I think)

Reply:
I'm a bit rusty on c++(no pun intended :)), but:

I used a similar approach although using `std::vector<InputPath>`, then it turned out that `InputPath` is itself a vector, so to avoid the copy I decided to switch to `std::vector<const InputPath&>`, but apparently c++ does not like vectors of vector references, which have to be wrapped in `std::reference_wrapper`. So I decided to convert to path strings for visited.

Like you mention, even using `vector<InputPath>` will likely not matter in terms of performance, switched back to that.

Reply:
Oh indeed, we're copying too much any way. Well, so be it then, we'll live with that.